cmake/soci_define_backend_target.cmake

```diff
@@ -123,7 +121,7 @@ function(soci_define_backend_target)
     target_link_libraries(${DEFINE_BACKEND_TARGET_NAME} PUBLIC "${CURRENT_DEP}")
   endforeach()
 
-  target_include_directories(${DEFINE_BACKEND_TARGET_NAME} PRIVATE ${DEFINE_BACKEND_PRIVATE_INCLUDE_DIRS})
+  target_include_directories(${DEFINE_BACKEND_TARGET_NAME} PRIVATE "${PROJECT_SOURCE_DIR}/include/private")
```

Comment:
I think being explicit when calling the function is better than doing this implicitly here. Otherwise, we get back to the CMake structure being somewhat harder to understand

Reply:
Sorry, I really don't see any difference between e.g. `target_include_directories(soci_core "${PROJECT_SOURCE_DIR}/include/soci")` and this line. Both define the include paths that should be used for all targets and I have trouble seeing how this one is more implicit than the other one.

OTOH I definitely do see a problem (not a huge one, but still) with repeating the same parameter unnecessarily in all function invocation. I don't know much about CMake, but you wouldn't have a function parameter whose value is always fixed in any other language, so why would you do it here?

Reply:
Actually, looking at this function more carefully, I see that it has another parameter, `REQUIRED_COMPONENTS`, which is not only always set to the same value, but can't possibly be ever set to anything else, so I've added a commit removing it too.

I've realized by now that we have very different ideas about how maintainable CMake code looks like and I definitely could be wrong, but I just don't see any possible rationale for using function arguments for something that is inherently fixed.

Reply:
There is a huge difference. The difference is in visibility. If you want to know how a given backend is set up, you check the CMakeLists.txt file for that backend. And in one version all include directory and dependencies are listed right there and in your version (same as before the CMake refactor) they are not. Hence, the reader of that code doesn't know about them without having to backtrack through the CMake code in order to find the definition of the backend definition function.

This is exactly the reason why the old CMake setup was so difficult to figure out. Everything was happening implicitly and you effectively had to read the entirety of the CMake code in order to be sure to know what is going on.

Reply:
I had never had this problem with the old CMake setup and I don't think I've heard anybody else (other than you, of course) complaining about it. The worst problem was the very confusing logic for enabling/detecting backends, but I had absolutely never worried about not being able to see that backend include directories contained `include/private` or that it was difficult to see that backends linked with "core" library and, frankly, I can't even imagine how could either of those be difficult to figure out.

Reply:
> but you wouldn't have a function parameter whose value is always fixed in any other language, so why would you do it here?

If the function in question was having global side-effects, you might want to keep apparently redundant parameters in there too, if those make those side-effects clear.

Besides, CMake is primarily a build system and not a real programming language per se. Hence, the support for tooling such as language servers etc. is just not there and hence you usually can't quickly jump to the definition of a given CMake macro/function. Thus, having things in-place makes it much easier to understand.

Reply:
Also, just for completeness: it's _vastly_ simpler to me to grep for `target_include_directories()` than to figure out that I have to look at `PRIVATE_INCLUDE_DIRS` parameter of the function (especially because it is being actually used as `DEFINE_BACKEND_PRIVATE_INCLUDE_DIRS`, so the usual `grep -w` wouldn't even find this!).

I think the only reason you think the latter is more understandable is that you wrote it and know how it works.

Reply:
Grepping is not really a valid approach (in general) to figure out a CMake setup as it misses context and might be returning matches for a completely unrelated target.

I would argue that it is much more natural to start the search for something like this in the CMake file for the component one is interested in...